### PR TITLE
enh(timeperiod): Replace date and time range by classes.

### PR DESCRIPTION
**Pull request by SylvestreG**
Replace date and time range by classes.

### CODE REVIEW

File: tests/timeperiod/utils.cc [hunk 1]
```diff
@@ -279,10 +295,15 @@ void timeperiod_creator::new_timerange(
                            int end_hour,
                            int end_minute,
                            daterange* target) {
-  add_timerange_to_daterange(
-    target,
-    hmtos(start_hour, start_minute),
-    hmtos(end_hour, end_minute));
+  if(!target)
```
Comment (bouda1) [inline tests/timeperiod/utils.cc]: boo
`if (!target)`

File: tests/timeperiod/utils.cc [hunk 2]
```diff
@@ -279,10 +295,15 @@ void timeperiod_creator::new_timerange(
...
+    return;
+
+  std::shared_ptr <timerange> tr =
+    std::make_shared<timerange>(
```
Comment (bouda1) [inline tests/timeperiod/utils.cc]: as previously...

File: tests/timeperiod/utils.cc [hunk 3]
```diff
@@ -303,13 +324,16 @@ void timeperiod_creator::new_timerange(
...
                            int day,
                            timeperiod* target) {
-  if (!target)
-    target = _timeperiods;
-  add_timerange_to_timeperiod(
-    target,
-    day,
-    hmtos(start_hour, start_minute),
-    hmtos(end_hour, end_minute));
+  if(!target)
```
Comment (bouda1) [inline tests/timeperiod/utils.cc]: if followed by space...

File: inc/com/centreon/engine/daterange.hh [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+CCE_END()
+
+typedef std::list<std::shared_ptr<com::centreon::engine::daterange>>
+  daterange_list;
+
+CCE_BEGIN()
+
+class daterange {
+ public:
+  daterange(
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply inc/com/centreon/engine/daterange.hh]: done

File: inc/com/centreon/engine/daterange.hh [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  int            _ewday_offset;
+  int            _skip_interval;
+};
+
+CCE_END()
+
+std::ostream&      operator<<(std::ostream& os, com::centreon::engine::daterange const& obj);
+
+#endif // !CCE_OBJECTS_DATERANGE_HH
+
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply inc/com/centreon/engine/daterange.hh]: done

File: inc/com/centreon/engine/hostdependency.hh [hunk 6]
```diff
@@ -89,13 +88,14 @@ class                           hostdependency {
 
   com::centreon::engine::host*  master_host_ptr;
   com::centreon::engine::host*  dependent_host_ptr;
-  timeperiod_struct*            dependency_period_ptr;
+  com::centreon::engine::timeperiod*
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply inc/com/centreon/engine/hostdependency.hh]: already fixed in next PR :)

File: inc/com/centreon/engine/timeperiod.hh [hunk 7]
```diff
@@ -1,6 +1,5 @@
 /*
-** Copyright 1999-2009      Ethan Galstad
-** Copyright 2011-2014,2016 Centreon
+** Copyright 2011-2013 Merethis
 **
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply inc/com/centreon/engine/timeperiod.hh]: a merge of timeperiod.h and objects/timperiod.h

It is fixed now.

File: inc/com/centreon/engine/timerange.hh [hunk 8]
```diff
@@ -0,0 +1,59 @@
+/*
+** Copyright 2011-2013 Merethis
+**
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply inc/com/centreon/engine/timerange.hh]: done

File: src/configuration/applier/timeperiod.cc [hunk 9]
```diff
@@ -306,23 +312,24 @@ void applier::timeperiod::_add_exceptions(
...
-          throw (engine_error()
-                 << "Could not add time range to date range of "
-                 << "type " << it2->type() << " of time period '"
-                 << tp->name << "'");
+           ++it3) {
+        std::shared_ptr<engine::timerange> tr =
+          std::make_shared<engine::timerange>(
+            it3->start(),
+            it3->end());
+        for(daterange_list::iterator
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply src/configuration/applier/timeperiod.cc]: OK.

File: src/configuration/applier/timeperiod.cc [hunk 10]
```diff
@@ -346,14 +353,13 @@ void applier::timeperiod::_add_time_ranges(
...
-             day,
-             it2->start(),
-             it2->end()))
-        throw (engine_error()
-               << "Could not add time range to time period '"
-               << tp->name << "'");
+         ++it2) {
+
+      std::shared_ptr<engine::timerange> tr =
+        std::make_shared<engine::timerange>(
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply src/configuration/applier/timeperiod.cc]: OK

File: src/daterange.cc [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+  int smon,
+  int smday,
+  int swday,
+  int swday_offset,
+  int eyear,
+  int emon,
+  int emday,
+  int ewday,
+  int ewday_offset,
+  int skip_interval) {
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply src/daterange.cc]: done

File: tests/timeperiod/utils.cc (hunk 1 above)
Comment (SylvestreG) [reply tests/timeperiod/utils.cc]: done

File: tests/timeperiod/utils.cc (hunk 2 above)
Comment (SylvestreG) [reply tests/timeperiod/utils.cc]: done

Comment (SylvestreG) [reply tests/timeperiod/utils.cc]: done

File: tests/timeperiod/utils.cc (hunk 3 above)
Comment (SylvestreG) [reply tests/timeperiod/utils.cc]: done

**Pull request merged**